### PR TITLE
Refactor Blockchain.Start 1787

### PR DESCRIPTION
**Pull request by lzxm160**
work on #1787

### CODE REVIEW

File: state/factory/factory.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ type (
 		SimulateExecution(context.Context, address.Address, *action.Execution, evm.GetBlockHash) ([]byte, *action.Receipt, error)
 		Commit(WorkingSet) error
 		State(hash.Hash160, interface{}) error
+		PutBlock(context.Context, []action.SealedEnvelope) error
```
Comment (lzxm160) [inline state/factory/factory.go]: There's import cycle error if using block.Block

File: blockchain/blockdao/blockdao.go [hunk 2]
```diff
@@ -120,6 +124,7 @@ type (
 		compressBlock bool
 		kvstore       db.KVStore
 		indexer       BlockIndexer
+		sf            factory.Factory
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: sf is just another indexer, don't use type factory.Factory

Comment (lzxm160) [reply blockchain/blockdao/blockdao.go]: you mean BlockIndexer interface?sf need change and add PutBlock,DeleteTipBlock,factory.Factory cannot use block.Block

**Comment by lizhefeng** [comment]
What is the progress of this PR?

**Comment by CoderZhi** [comment]
> What is the progress of this PR?

@koseoyoung is working on a task to move working set out of block. After that, we can resume this PR.

**Comment by koseoyoung** [comment]
we can resume this PR

**Comment by lzxm160** [comment]
restart this pr today

**Comment by lzxm160** [comment]
After #1864 and #1863,this will be resume and test again

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -290,17 +297,7 @@ func (bc *blockchain) Start(ctx context.Context) error {
...
-		for _, p := range bcCtx.Registry.All() {
-			if s, ok := p.(lifecycle.Starter); ok {
-				if err := s.Start(ctx); err != nil {
-					return errors.Wrap(err, "failed to start protocol")
-				}
-			}
-		}
-	}
-
-	return bc.startExistingBlockchain(ctx)
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: why do we still keep https://github.com/iotexproject/iotex-core/pull/1819/files#diff-646235a88d91f225fc74104cca24062aL513-R546 ?

File: blockchain/blockdao/blockdao.go [hunk 4]
```diff
@@ -134,14 +141,43 @@ type (
...
+		if dao.indexer != nil {
+			return nil
+		}
+		dao.indexer = indexer
+		return nil
+	}
+}
+
+// FactoryOption sets dao with BlockIndexer
+func FactoryOption(sf factory.Factory) Option {
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: sf BlockIndexer

File: blockchain/blockdao/blockdao.go [hunk 5]
```diff
@@ -105,21 +110,23 @@ type (
...
+		PutBlock(ctx context.Context, blk *block.Block) error
 		DeleteTipBlock(blk *block.Block) error
-		Commit() error
 	}
 
 	blockDAO struct {
 		compressBlock bool
 		kvstore       db.KVStore
 		indexer       BlockIndexer
+		sf            BlockIndexer
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: sf is just another indexer, why do you distinguish it?

File: blockchain/blockdao/blockdao.go [hunk 6]
```diff
@@ -158,8 +194,8 @@ func NewBlockDAO(kvstore db.KVStore, indexer BlockIndexer, compressBlock bool, c
 	}
 	blockDAO.timerFactory = timerFactory
 	blockDAO.lifecycle.Add(kvstore)
-	if indexer != nil {
-		blockDAO.lifecycle.Add(indexer)
+	if blockDAO.indexer != nil {
+		blockDAO.lifecycle.Add(blockDAO.indexer)
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: what about sf?

File: blockchain/blockdao/blockdao.go [hunk 7]
```diff
@@ -216,6 +252,43 @@ func (dao *blockDAO) initStores() error {
...
+		}
+		producer, err := address.FromBytes(blk.PublicKey().Hash())
+		if err != nil {
+			return err
+		}
+		ctx = dao.contextWithBlock(ctx, producer, blk.Height(), blk.Timestamp(), gas)
+		if err := dao.sf.PutBlock(ctx, blk); err != nil {
+			return err
+		}
+	}
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: this is something sf should do by itself

File: state/factory/factory.go [hunk 8]
```diff
@@ -328,6 +330,48 @@ func (sf *factory) Commit(ctx context.Context, blk *block.Block) error {
 	timer := sf.timerFactory.NewTimer("Commit")
 	sf.mutex.Unlock()
 	defer timer.End()
+	return sf.commitBlock(ctx, blk)
+}
+
+// PutBlock persists all changes in RunActions() into the DB
+func (sf *factory) PutBlock(ctx context.Context, blk *block.Block) error {
```
Comment (CoderZhi) [inline state/factory/factory.go]: If you introduce this function, delete Commit

File: state/factory/factory.go [hunk 9]
```diff
@@ -328,6 +330,48 @@ func (sf *factory) Commit(ctx context.Context, blk *block.Block) error {
...
+	return sf.commitBlock(ctx, blk)
+}
+
+// DeleteTipBlock delete blk
+func (sf *factory) DeleteTipBlock(blk *block.Block) error {
+	return nil
```
Comment (CoderZhi) [inline state/factory/factory.go]: throw an error

File: state/factory/statedb.go [hunk 10]
```diff
@@ -229,6 +229,54 @@ func (sdb *stateDB) Commit(ctx context.Context, blk *block.Block) error {
...
+
+// Commit persists all changes in RunActions() into the DB
+func (sdb *stateDB) PutBlock(ctx context.Context, blk *block.Block) error {
+	return sdb.commitBlock(ctx, blk)
+}
+
+// DeleteTipBlock delete blk
+func (sdb *stateDB) DeleteTipBlock(blk *block.Block) error {
+	return nil
+}
```
Comment (CoderZhi) [inline state/factory/statedb.go]: same here

**Comment by CoderZhi** [comment]
Please move on some other tasks first if you got stuck in this PR.

**Comment by lzxm160** [comment]
> Please move on some other tasks first if you got stuck in this PR.

sure

File: blockchain/blockchain.go (hunk 3 above)
Comment (lzxm160) [reply blockchain/blockchain.go]: my fault,forgot to delete when merge

File: blockchain/blockdao/blockdao.go (hunk 7 above)
Comment (lzxm160) [reply blockchain/blockdao/blockdao.go]: have to copy all GetBlockByHeight related function to sf,I think it's not necessary

**Comment by lzxm160** [comment]
Temporarily close, and reopen if necessary

**Pull request closed without merging**